gpbft/gpbft.go

```diff
@@ -918,6 +948,66 @@ func (i *instance) broadcast(round uint64, step Phase, value ECChain, createTick
 	}
 
 	_ = i.participant.host.RequestBroadcast(mb)
+	i.broadcasted.record(mb)
+}
+
+// tryRebroadcast checks whether the conditions for rebroadcasting messages are
+// met, and if so rebroadcasts messages from current and previous rounds.
+// Rebroadcast is needed when both of the following conditions are met:
+//   - There has been no progress since latest broadcast.
+//   - Rebroadcast backoff duration has elapsed.
+//
+// If there has been progress since latest broadcast this function resets the
+// number of rebroadcast attempts and returns immediately.
+func (i *instance) tryRebroadcast() error {
+	if i.progressedSinceLatestBroadcast() {
+		i.rebroadcastAttempts = 0
+		return nil
+	}
+
+	if i.rebroadcastAttempts == 0 {
+		// It is the first time that rebroadcast is necessary; set initial rebroadcast timeout.
+		i.rebroadcastTimeout = i.phaseTimeout.Add(i.participant.rebroadcastAfter(0))
+	}
+
+	if i.rebroadcastTimeoutElapsed() {
+		err := i.rebroadcast()
+		i.rebroadcastAttempts++
+		i.rebroadcastTimeout = i.rebroadcastTimeout.Add(i.participant.rebroadcastAfter(i.rebroadcastAttempts))
+		return err
+	}
+
+	// Schedule the next rebroadcast.
+	i.participant.host.SetAlarm(i.rebroadcastTimeout)
+	return nil
+}
+
+func (i *instance) rebroadcastTimeoutElapsed() bool {
+	now := i.participant.host.Time()
+	return atOrAfter(now, i.rebroadcastTimeout)
+}
+
+func (i *instance) progressedSinceLatestBroadcast() bool {
+	return i.phase != i.broadcasted.phaseAtLatestBroadcast
+}
+
+func (i *instance) rebroadcast() error {
+
+	// Rebroadcast all messages from the current and previous rounds, unless the
+	// instance has progressed to DECIDE phase. In which case, only DECIDE message is
+	// rebroadcasted.
+	//
+	// Note that the implementation here rebroadcasts more messages than FIP-0086
+	// strictly requires. Because, the cost of rebroadcasting additional messages is
+	// small compared to the reduction in need for rebroadcast.
+	for _, mbs := range i.broadcasted.messagesByRound {
```

Comment:
We can optimise further here by sorting messages before broadcast. But that is throw of the dice optimisation since there are no guaranteed ordering in message broadcast. Hence the simpler approach here.